File: apps/web/src/lib/types/product.ts
import type { EnvironmentType, ProjectSourceType } from "./enums";

export interface Product {
  id: string;
  name: string;
  status: string | null;
  stage: string | null;
  pillar: string | null;
  progress: number | null;
  health_score: number | null;
  logo_url: string | null;
  repository_url: string | null;
  lovable_url: string | null;
  source_type: ProjectSourceType | null;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProfileSummary {
  id: string;
  full_name: string | null;
  email: string | null;
  avatar_url: string | null;
}

export interface ProductMember {
  id: string;
  product_id: string;
  profile_id: string;
  role: string | null;
  created_at: string;
  profile: ProfileSummary | null;
}

export interface ProductEnvironment {
  id: string;
  product_id: string;
  environment_type: EnvironmentType;
  url: string | null;
  branch: string | null;
  status: string;
  target_domain: string | null;
  railway_project_url: string | null;
  railway_toml_present: boolean | null;
  last_deployment_at: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export type ProductMemberRole = "project_manager" | "marketing" | "business_owner" | "ai_engineer";

export interface TeamReadiness {
  complete: boolean;
  members: ProductMember[];
  missing: string[];
}

export interface ProductDocument {
  id: string;
  product_id: string;
  file_name: string;
  file_path: string;
  file_size: number;
  file_type: string;
  category: string | null;
  description: string | null;
  ai_summary: string | null;
  summary_generated_at: string | null;
  folder_id: string | null;
  uploaded_by: string;
  created_at: string;
  updated_at: string;
}

File: apps/api/models/product.py
"""Product-related models: products, members, environments, documents, notes."""

import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from packages.common.db.base import Base, TimestampMixin, UUIDMixin


class Product(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "products"

    name: Mapped[str] = mapped_column(String, nullable=False)
    status: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    stage: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    pillar: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    progress: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    health_score: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    repository_url: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    lovable_url: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    logo_url: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    source_type: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    archived_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )


class ProductMember(Base, UUIDMixin):
    __tablename__ = "product_members"

    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id"), nullable=False
    )
    profile_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("profiles.id"), nullable=False
    )
    role: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    profile: Mapped["Profile"] = relationship(
        "Profile", foreign_keys=[profile_id], lazy="selectin"
    )


class ProductEnvironment(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "product_environments"

    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id"), nullable=False
    )
    environment_type: Mapped[str] = mapped_column(String, nullable=False)
    url: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    branch: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    status: Mapped[str] = mapped_column(String, default="active")
    target_domain: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    railway_project_url: Mapped[Optional[str]] = mapped_column(
        String, nullable=True
    )
    railway_toml_present: Mapped[Optional[bool]] = mapped_column(
        Boolean, nullable=True
    )
    notes: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    last_deployment_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )


class ProductDocument(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "product_documents"

    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id"), nullable=False
    )
    uploaded_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("profiles.id"), nullable=False
    )
    folder_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("document_folders.id"), nullable=True
    )
    file_name: Mapped[str] = mapped_column(String, nullable=False)
    file_path: Mapped[str] = mapped_column(String, nullable=False)
    file_type: Mapped[str] = mapped_column(String, nullable=False)
    file_size: Mapped[int] = mapped_column(Integer, nullable=False)
    category: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    description: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    ai_summary: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    summary_generated_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )


class ProductManagementNote(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "product_management_notes"

    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id"), nullable=False
    )
    author_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("profiles.id"), nullable=False
    )
    content: Mapped[str] = mapped_column(String, nullable=False)
    is_pinned: Mapped[bool] = mapped_column(Boolean, default=False)


class ProductPartnerNote(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "product_partner_notes"

    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id"), nullable=False
    )
    author_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("profiles.id"), nullable=False
    )
    partner_name: Mapped[str] = mapped_column(String, nullable=False)
    content: Mapped[str] = mapped_column(String, nullable=False)


class ProductNotificationSetting(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "product_notification_settings"

    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("products.id", ondelete="CASCADE"),
        unique=True,
        nullable=False,
    )
    email_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
